components/youtube_script_injector/browser/core/youtube_registry.h

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_COMPONENTS_YOUTUBE_SCRIPT_INJECTOR_BROWSER_CORE_YOUTUBE_REGISTRY_H_
+#define BRAVE_COMPONENTS_YOUTUBE_SCRIPT_INJECTOR_BROWSER_CORE_YOUTUBE_REGISTRY_H_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "base/component_export.h"
+#include "base/gtest_prod_util.h"
+#include "base/memory/raw_ptr.h"
+#include "base/memory/singleton.h"
+#include "base/memory/weak_ptr.h"
+#include "brave/components/youtube_script_injector/browser/core/youtube_json.h"
+
+class GURL;
+
+namespace youtube_script_injector {
+// Needed for testing private methods in YouTubeTabHelperBrowserTest.
+FORWARD_DECLARE_TEST(YouTubeTabHelperBrowserTest, NoMatch);
+FORWARD_DECLARE_TEST(YouTubeTabHelperBrowserTest, RuleMatchTestScriptFalse);
+FORWARD_DECLARE_TEST(YouTubeTabHelperBrowserTest, RuleMatchTestScriptTrue);
+
+// This class loads and stores the rules from the youtube.json file.
```

Comment:
Maybe we could have more documentation here. What are some example rules? Is the JSON part of this PR? If not, maybe an example ruleset could be useful.

Reply:
OK I think I have a partial idea of what this is supposed to do. This should be genericizing any scripting that we want to do for YouTube. For example, we previously had logic in [background_video_playback_tab_helper.cc](https://github.com/brave/brave-core/blob/e26734efeb0a24b767e32741517b1f45ca4a38a9/browser/android/background_video/background_video_playback_tab_helper.cc#L22-L32) and this PR is removing that file. The background play is one example script that gets loaded.

Where does it get loaded from?

Reply:
OK seems to get loaded from component updater 😄 I see the comments in `youtube_component_installer.cc` which has the example directory structure and then `youtube_json.h` has an example of the JSON.

After finding it out maybe just a comment like:

```
- See youtube_json.h for an example of the JSON
- See youtube_component_installer.cc for example of the component layout
```

Reply:
> This should be genericizing and centralizing any scripting that we want to do for YouTube, right?

Correct! That's the idea. But I haven't created the extension (yet) I've temporarily used extension ID (lhhcaamjbmbijmjbnnodjaknblkiagon) PSST. I'm going over https://github.com/brave/brave-core-crx-packager now. Probably I won't be able to complete the full process as presumably I don't have full access. I'll post an update soon.

> Maybe we could have more documentation here. What are some example rules? Is the JSON part of this PR? If not, maybe an example ruleset could be useful.

The idea that seems the best compromize to me would be to have some tests that contain the JS we are injecting.
Same as done by Shivan here https://github.com/brave/brave-core/tree/master/test/data/psst-component-data

Reply:
Improved code comments 👉 [`ae6fac2` (#26355)](https://github.com/brave/brave-core/pull/26355/commits/ae6fac23b9f9e891540a7833fbd695f6eca9842c)